dashpay/src/main/kotlin/org/dashj/platform/dashpay/BlockchainIdentity.kt

```diff
@@ -842,49 +867,59 @@ class BlockchainIdentity {
         return privateKey
     }
 
+    fun getIdentityPublicKeyByPurpose(purpose: KeyIndexPurpose): IdentityPublicKey {
+        return identity!!.getPublicKeyById(purpose.ordinal)!!
+    }
+
     fun signStateTransition(transition: StateTransitionIdentitySigned, keyParameter: KeyParameter?) {
         checkIdentity()
+        val identityPublicKey = getIdentityPublicKeyByPurpose(KeyIndexPurpose.AUTHENTICATION)
         return signStateTransition(
             transition,
-            identity!!.publicKeys[0].id/* currentMainKeyIndex*/,
-            currentMainKeyType,
+            KeyIndexPurpose.AUTHENTICATION.ordinal,
+            identityPublicKey.type,
             keyParameter
         )
     }
 
-    fun derivationPathForType(type: IdentityPublicKey.TYPES): ImmutableList<ChildNumber>? {
+    fun derivationPathForType(type: IdentityPublicKey.Type): ImmutableList<ChildNumber>? {
         if (isLocal) {
-            if (type == IdentityPublicKey.TYPES.ECDSA_SECP256K1) {
+            if (type == IdentityPublicKey.Type.ECDSA_SECP256K1) {
                 return DerivationPathFactory(wallet!!.params).blockchainIdentityECDSADerivationPath()
-            } else if (type == IdentityPublicKey.TYPES.BLS12_381) {
+            } else if (type == IdentityPublicKey.Type.BLS12_381) {
                 return DerivationPathFactory(wallet!!.params).blockchainIdentityBLSDerivationPath()
             }
         }
         return null
     }
 
-    private fun privateKeyAtIndex(index: Int, type: IdentityPublicKey.TYPES): ECKey? {
-        Preconditions.checkState(isLocal, "this must own a wallet")
+    /**
+     * Get a decrypted private key at index for the specified key type
+     *
+     * @param index the index of the key to obtain see [KeyIndexPurpose]
+     * @param type the type of key to obtain
+     * @param keyParameter the encryption key of the wallet
+     * @return the key that matches the input parameters or null
+     */
+    private fun privateKeyAtIndex(index: Int, type: IdentityPublicKey.Type, keyParameter: KeyParameter?): ECKey? {
+        checkState(isLocal, "this must own a wallet")
 
         when (type) {
-            IdentityPublicKey.TYPES.ECDSA_SECP256K1 -> {
+            IdentityPublicKey.Type.ECDSA_SECP256K1 -> {
                 val authenticationChain = wallet!!.blockchainIdentityKeyChain
-                val key = authenticationChain.watchingKey
+                // decrypt keychain
+                val decryptedChain = if (wallet!!.isEncrypted) {
+                    authenticationChain.toDecrypted(keyParameter)
+                } else {
+                    authenticationChain
+                }
+                val key = decryptedChain.getKey(index) // watchingKey
+                checkState(key.path.last().isHardened)
```

Comment:
This is probably slow - we are decrypting the key chain every time we need a key which also requires deriving the key we need instead of decrypting only the key that we need.

But the keychain doesn't yet contain the keys we use for different purposes.  Instead, it contains only the parent key from which a master level key and a high-level key will be derived each time it is needed.
```
blockchain identity keys derivation path (EC Keys)

m/9'/1'/5'/0'/0'/0'    - key chain (parent)
m/9'/1'/5'/0'/0'/0'/0' - Master Key
m/9'/1'/5'/0'/0'/0'/1' - High Key


```

In Platform 0.22, we only had 1 key to keep track of and it was already included in the key chain.

Also, we haven't defined a standard key creation scheme with iOS.  Platform requires 2 keys with Master and High security levels.